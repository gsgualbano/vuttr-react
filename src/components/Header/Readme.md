#### Cabeçalho da página, onde ficam título, subtítulo, botão para adicionar nova ferramenta, botão para mudar tema e caixa de pesquisa.

Exemplo:

```js
<Header />
```
