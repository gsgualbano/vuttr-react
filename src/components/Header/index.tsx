import React, { useState } from 'react';

import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';

import * as ToolsAction from '~/store/ducks/tools/actions';

import useDarkMode from 'use-dark-mode';

import Modal from '../Modal';
import AddForm from '~/components/Forms/Add';
import PrimaryButton from '~/components/Common/PrimaryButton';

import {
  Container,
  Title,
  SubTitle,
  Row,
  Search,
  SearchInput,
  SearchIcon,
  Input,
  CheckBox,
  IconContainer,
  PlusIcon,
  ThemeButton,
  DarkIcon,
  LightIcon,
} from './styles';

/**
 * Props que vem do redux
 */
interface DispatchProps {
  /** ação para pesquisar ferramentas */
  searchToolGeneral(data: string): void;
  /** ação para pesquisar ferramentas por tag */
  searchToolTag(data: string): void;
}

/**
 * Cabeçalho da página
 */
const Header = ({ searchToolGeneral, searchToolTag }: DispatchProps): JSX.Element => {
  // custom hook that creates a state shared across components and saves on local storage
  const darkMode = useDarkMode();

  // states
  const [showModal, setShowModal] = useState<boolean>(false);
  const [checkbox, setCheckBox] = useState<boolean>(true);

  // show modal when add button pressed
  const handleAddPress = (e: React.MouseEvent<HTMLElement>): void => {
    setShowModal(true);
  };

  // Input change of the search box
  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    if (checkbox) {
      return searchToolTag(e.target.value); 
    }

    return searchToolGeneral(e.target.value);
  };

  // On the check or discheck of the checkbox
  const handleCheckBoxChange = (e: React.ChangeEvent<HTMLElement>): void => {
    setCheckBox(!checkbox);
  };

  return (
    <Container>
      <Row>
        <Title>VUTTR</Title>
        <ThemeButton type="button" onClick={darkMode.toggle}>
          {darkMode.value ? (
            <>
              <LightIcon />
              Light Theme
            </>
          ) : (
            <>
              <DarkIcon />
              Dark Theme
            </>
          )}
        </ThemeButton>
      </Row>
      <SubTitle>Very Useful Tools to Remember</SubTitle>
      <Row>
        <Search>
          <SearchInput>
            <SearchIcon />
            <Input placeholder="Search" onChange={handleInputChange} />
          </SearchInput>
          <CheckBox defaultChecked={checkbox} onChange={handleCheckBoxChange} />
          search in tags only
        </Search>
        <PrimaryButton onClick={handleAddPress}>
          <IconContainer>
            <PlusIcon />
          </IconContainer>
          Add
        </PrimaryButton>
      </Row>
      <Modal visible={showModal} setShowModal={setShowModal}>
        <AddForm setShowModal={setShowModal} />
      </Modal>
    </Container>
  );
};

const mapDispatchToProps = (dispatch: Dispatch) => bindActionCreators(ToolsAction, dispatch);

export default connect(
  null,
  mapDispatchToProps,
)(Header);
