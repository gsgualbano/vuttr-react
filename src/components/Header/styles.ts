import styled from 'styled-components';
import { ReactComponent as PIcon } from '~/assets/icons/Icon-Plus-Circle-2px.svg';
import { ReactComponent as SIcon } from '~/assets/icons/Icon-Search-2px.svg';
import { ReactComponent as MoonIcon } from '~/assets/icons/Icon-Dark Mode 2px.svg';
import { ReactComponent as SunIcon } from '~/assets/icons/Icon-Light-Mode-2px.svg';

export const Container = styled.div``;

export const Title = styled.h1`
  font-weight: 600;
  font-size: 42px;
  margin: 30px 0;
  letter-spacing: 2px;
`;

export const SubTitle = styled.h2`
  font-weight: 600;
  font-size: 30px;
  margin: 30px 0;
  letter-spacing: 2px;
`;

export const Row = styled.div`
  display: flex;
  flex-direction: row;
  flex-flow: row wrap;
  justify-content: space-between;
  margin: 10px 0;
`;

export const Search = styled.div`
  display: flex;
  flex-direction: row;
  flex-flow: row wrap;
  justify-content: center;
  align-items: center;

  @media (max-width: 768px) {
    margin-bottom: 10px;
    justify-content: none;
  }
`;

export const SearchInput = styled.div`
  display: flex;
  flex-direction: row;
  border-radius: 5px;
  border: 1px solid;
  border-color: ${({ theme }) => theme.colors.InputBorderColor};
  background-color: ${({ theme }) => theme.colors.InputBackground};
  width: 200pt;
  margin-right: 10px;
  padding: 10px;
  font-family: 'Source Sans Pro', sans-serif;

  @media (max-width: 767px) {
    width: 100%;
  }
`;

export const Input = styled.input.attrs({
  type: 'text',
})`
  border: none;
  width: 100%;
  background: inherit;
  outline: none;
`;

export const CheckBox = styled.input.attrs({
  type: 'checkbox',
})``;

export const IconContainer = styled.div`
  display: flex;
  height: 100%;
  justify-content: center;
  align-items: center;
`;

export const PlusIcon = styled(PIcon)`
  width: 14px;
  height: 14px;
  margin-right: 5px;

  g > circle.a {
    stroke: #ffffff;
    stroke-width: 2px;
  }

  g > path.b {
    stroke: #ffffff;
    stroke-width: 7px;
  }
`;

export const SearchIcon = styled(SIcon)`
  width: 14px;
  height: 14px;
  margin-right: 5px;

  g > path.a {
    stroke: #b1adb9;
  }

  g > circle.a {
    stroke: #b1adb9;
  }
`;

export const ThemeButton = styled.button`
  max-height: 30px;
  background: transparent;
  border: none;
  color: #b1adb9;
  font-size: 20px;
  margin: auto 0;
`;

export const DarkIcon = styled(MoonIcon)`
  width: 18px;
  height: 18px;
  margin-right: 7px;

  g > path.a {
    stroke: #b1adb9;
  }

  g > circle.a {
    stroke: #b1adb9;
  }
`;

export const LightIcon = styled(SunIcon)`
  width: 14px;
  height: 14px;
  margin-right: 5px;

  g > path.a {
    stroke: #b1adb9;
  }

  g > circle.a {
    stroke: #b1adb9;
  }
`;
