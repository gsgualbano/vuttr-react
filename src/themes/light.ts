export default {
  colors: {
    background: '#fcfcfd',
    textColor: '#170c3a',
    cardBackground: '#fcfcfd',
    cardBorderColor: '#ebeaed',
    modalBackground: '#fefefe',
    modalBorderColor: '#888',
    linkColor: 'rgb(0, 0, 238)',
    linkVisitedColor: 'rgb(85, 26, 139);',
    buttonOutlineBackground: '#ffffff',
    InputBackground: '#f5f4f6',
    InputBorderColor: '#ebeaed',
  },
};
