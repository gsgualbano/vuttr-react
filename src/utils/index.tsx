/* eslint-disable max-len */
import React from 'react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import { render } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { reducer } from '../store/ducks/tools';

interface Theme {
  colors: {
    background: string;
    textColor: string;
    cardBackground: string;
    cardBorderColor: string;
    modalBackground: string;
    modalBorderColor: string;
    linkColor: string;
    linkVisitedColor: string;
    buttonOutlineBackground: string;
    InputBackground: string;
    InputBorderColor: string;
  };
}

/** redux mock to render components that depends on the redux */
export const renderWithRedux = (
  component: JSX.Element,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(<Provider store={store}>{component}</Provider>),
  store,
});

/** render components thats depends on the theme provider */
export const renderWithTheme = (component: JSX.Element, theme: Theme) => render(<ThemeProvider theme={theme}>{component}</ThemeProvider>);

/** render components thats depends on the redux and theme provider */
export const renderWithReduxAndTheme = (
  component: JSX.Element,
  theme: Theme,
  { initialState, store = createStore(reducer, initialState) }: any = {},
) => ({
  ...render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>{component}</Provider>
    </ThemeProvider>,
  ),
  store,
});
