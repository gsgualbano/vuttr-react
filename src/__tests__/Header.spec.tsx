import React from 'react';
import { renderWithReduxAndTheme } from '~/utils';
import { cleanup, fireEvent } from '@testing-library/react';

import Header from '~/components/Header';

import 'jest-dom/extend-expect';

import light from '~/themes/light';

describe('<Header />', () => {
  afterEach(cleanup);

  it('render correctly', () => {
    const { asFragment } = renderWithReduxAndTheme(<Header />, light);
    expect(asFragment()).toMatchSnapshot();
  });

  it('show modal', () => {
    const { getByText } = renderWithReduxAndTheme(<Header />, light);
    fireEvent.click(getByText('Add'));
    expect(getByText('Add new tool')).toBeInTheDocument();
    expect(getByText('Tool Name')).toBeInTheDocument();
    expect(getByText('Tool Link')).toBeInTheDocument();
    expect(getByText('Tool description')).toBeInTheDocument();
    expect(getByText('Tags')).toBeInTheDocument();
  });

  it('Closes modal on ouside click', () => {
    const { queryByText, getByText, getByTestId } = renderWithReduxAndTheme(<Header />, light);
    fireEvent.click(getByText('Add'));
    fireEvent.click(getByTestId('outside-modal'));
    expect(queryByText('Add new tool')).not.toBeInTheDocument();
    expect(queryByText('Tool Name')).not.toBeInTheDocument();
    expect(queryByText('Tool Link')).not.toBeInTheDocument();
    expect(queryByText('Tool description')).not.toBeInTheDocument();
    expect(queryByText('Tags')).not.toBeInTheDocument();
  });
});
