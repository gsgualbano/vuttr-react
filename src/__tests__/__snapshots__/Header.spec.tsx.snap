// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-kgoBCf kVSTZj"
  >
    <div
      class="sc-dxgOiQ cvlRKR"
    >
      <h1
        class="sc-kGXeez egQvhw"
      >
        VUTTR
      </h1>
      <button
        class="sc-hSdWYo eYuGPD"
        type="button"
      >
        <svg
          class="sc-eHgmQL ZGrFG"
        >
          Icon-Dark Mode 2px.svg
        </svg>
        Dark Theme
      </button>
    </div>
    <h2
      class="sc-kpOJdX heUMOI"
    >
      Very Useful Tools to Remember
    </h2>
    <div
      class="sc-dxgOiQ cvlRKR"
    >
      <div
        class="sc-ckVGcZ iywKtY"
      >
        <div
          class="sc-jKJlTe fRBAMh"
        >
          <svg
            class="sc-iAyFgw eXuNiC"
          >
            Icon-Search-2px.svg
          </svg>
          <input
            class="sc-eNQAEJ bJNCjU"
            placeholder="Search"
            type="text"
          />
        </div>
        <input
          checked=""
          class="sc-hMqMXs cPxEnv"
          type="checkbox"
        />
        search in tags only
      </div>
      <button
        class="sc-chPdSV iFkdLK"
        data-testid="primary-button"
        type="button"
      >
        <div
          class="sc-kEYyzF hhXvCW"
        >
          <svg
            class="sc-kkGfuU hFQqd"
          >
            Icon-Plus-Circle-2px.svg
          </svg>
        </div>
        Add
      </button>
    </div>
  </div>
</DocumentFragment>
`;
